sonic-xcvrd/xcvrd/sff_mgr.py

```diff
@@ -0,0 +1,407 @@
+                    # Skip if XcvrApi is not supported
+                    api = sfp.get_xcvr_api()
+                    if api is None:
+                        self.log_error(
+                            "{}: skipping sff_mgr since no xcvr api!".format(lport))
+                        continue
+
+                    # Skip if it's not a paged memory device
+                    if api.is_flat_memory():
+                        self.log_notice(
+                            "{}: skipping sff_mgr for flat memory xcvr".format(lport))
+                        continue
+
+                    # Skip if it's a copper cable
+                    if api.is_copper():
+                        self.log_notice(
+                            "{}: skipping sff_mgr for copper cable".format(lport))
+                        continue
+
+                    # Skip if tx_disable action is not supported for this xcvr
+                    if not api.get_tx_disable_support():
+                        self.log_notice(
+                            "{}: skipping sff_mgr due to tx_disable not supported".format(
+                                lport))
+                        continue
+                except AttributeError:
+                    # Skip if these essential routines are not available
+                    continue
+
+                # Only turn on TX if both host_tx_ready is true and admin_status is up
+                target_tx_disable_flag = not (data['host_tx_ready'] == 'true'
+                                              and data['admin_status'] == 'up')
+                # get_tx_disable API returns an array of bool, with tx_disable flag on each channel.
+                # True means tx disabled; False means tx enabled.
+                cur_tx_disable_array = api.get_tx_disable()
+                if cur_tx_disable_array is None:
+                    self.log_error("{}: Failed to get current tx_disable value".format(lport))
+                    # If reading current tx_disable/enable value failed (could be due to
+                    # read error), then set this variable to the opposite value of
+                    # target_tx_disable_flag, to let detla array to be True on
+                    # all the interested channels, to try best-effort TX disable/enable.
+                    cur_tx_disable_array = [not target_tx_disable_flag] * self.DEFAULT_NUM_CHANNELS
+                # Get an array of bool, where it's True only on the channels that need change.
+                delta_array = self.calculate_tx_disable_delta_array(cur_tx_disable_array,
+                                                                    target_tx_disable_flag, channel)
+                mask = self.convert_bool_array_to_bit_mask(delta_array)
+                if mask == 0:
+                    self.log_notice("{}: No change is needed for tx_disable value".format(lport))
+                    continue
+                if api.tx_disable_channel(mask, target_tx_disable_flag):
```

Comment:
Should we bring a module to high-power before enabling TX ? In CMIS State machine it is done as a part of CMIS Data Path SM

Reply:
For now, non-cmis/ccmis optics are brought to High Power mode by default by platform.
The missing part is: enabling high power class by setting page 0 byte 93 bit 2(if advertised power class >=5). Yes, this handling will be added to sff_mgr in separate PR.

Reply:
raised issue to track high power handling https://github.com/sonic-net/sonic-platform-daemons/issues/414